src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Join.cs

```diff
@@ -0,0 +1,244 @@
+                                valid = false;
+                                break;
+                            }
+
+                            rightTableClone = rightTableClone.Drop(ref fError, DPath.Root, dottedNameNode.Right.Name);
+                            rightTableClone = rightTableClone.Add(ref fError, DPath.Root, asNode.Right.Name, rightType);
+                            break;
+
+                        default:
+                            // No new error message. Binder already added error for unknown scope.
+                            valid = false;
+                            break;
+                    }
+
+                    if (fError)
+                    {
+                        valid = false;
+                        break;
+                    }
+                }
+            }
+
+            if (leftTableClone.CanUnionWithForcedUniqueColumns(rightTableClone, useLegacyDateTimeAccepts: false, features: context.Features, out var duplicatedDType))
+            {
+                returnType = DType.Union(leftTableClone, rightTableClone, false, Features.PowerFxV1);
+            }
+            else
+            {
+                errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrJoinCantUnion, string.Join(", ", duplicatedDType.GetAllNames(DPath.Root).Select(tname => tname.Name.Value)));
+                valid = false;
+            }
+            
+            return valid;
+        }
+
+        public override bool IsLambdaParam(TexlNode node, int index)
+        {
+            return index == 2 || (index > 3 && node is AsNode);
+        }
+
+        public override IEnumerable<string> GetRequiredEnumNames()
+        {
+            return new List<string>() { BuiltInEnums.JoinTypeEnum.EntityName.Value };
+        }
+
+        public override ParamIdentifierStatus GetIdentifierParamStatus(TexlNode node, Features features, int index)
+        {
+            return ParamIdentifierStatus.NeverIdentifier;
+        }
+
+        internal override IntermediateNode CreateIRCallNode(PowerFx.Syntax.CallNode node, IRTranslatorContext context, List<IntermediateNode> args, ScopeSymbol scope)
+        {
+            var carg = node.Args.Count;
+            var newArgs = new List<IntermediateNode>();
+            var recordTypesMap = new Dictionary<DName, RecordType>();
+            var recordValueMap = new Dictionary<DName, Dictionary<DName, IntermediateNode>>();
+            var sourceNameProvider = new Dictionary<DName, DName>();
+
+            for (int i = 0; i < carg; i++)
+            {
+                var arg = node.Args.Children[i];
+
+                // Left
+                if (i == 0)
+                {
+                    sourceNameProvider[arg is AsNode asNode ? asNode.Left.AsDottedName().Left.AsFirstName().Ident.Name : FunctionJoinScopeInfo.LeftRecord] = FunctionJoinScopeInfo.LeftRecord;
```

Comment:
>arg is AsNode asNode ? asNode.Left.AsDottedName().Left.AsFirstName().Ident.Name : FunctionJoinScopeInfo.LeftRecord [](http://example.com/codeflow?start=39&length=114)

nit - for readability, can we hoist this to a separate expression so that it's easier to read? 

ie, prefer a few shorter lines (with intermiedate assignments) over 1 larger line.